File: crates/types/src/boundary.rs
//! Block boundary markers. These do not carry the state root.
//!
//! Both markers carry the block's L1 origin. This is the L1 block number for
//! the epoch of this L2 block. This makes deposit derivation a pure function
//! of L1: a reconstructor that reads only L1 and the DA payload can find
//! which epoch's deposits go at the front of each block. See
//! `docs/agents/l1-origin-deposit-derivation-spec.md`.
//!
//! The sealer assigns the origin. The sealer never reads L1 itself. It copies
//! the value from an ordered origin-advancing record (see
//! `kardamom_cluster_adapter::wire::KIND_ORIGIN_RECORD`). Keeping L1 out of
//! the Raft state machine keeps the state machine deterministic across replicas.

use rkyv::{Archive, Deserialize, Serialize};

use crate::position::BPosition;

/// Block-boundary marker emitted by the sealer onto tx_ordering.
#[derive(Clone, Debug, Default, Eq, PartialEq, Archive, Serialize, Deserialize)]
#[rkyv(derive(Debug))]
pub struct BlockBoundaryStart {
    pub block_number: u64,
    pub end_tx_idx: BPosition,
    pub l2_timestamp: u64,
    /// L1 block number for this block's epoch. This is `0` until the first
    /// origin-advancing record is ordered, and on older chains from before such records.
    pub l1_origin: u64,
}

/// Block-boundary closeout. Executors emit this onto tx_receipts after they
/// finish executing through `end_tx_idx`. This struct has no `state_root_commitment` field.
#[derive(Clone, Debug, Default, Eq, PartialEq, Archive, Serialize, Deserialize)]
#[rkyv(derive(Debug))]
pub struct BlockBoundary {
    pub block_number: u64,
    pub end_tx_idx: BPosition,
    pub l2_timestamp: u64,
    /// This value comes from [`BlockBoundaryStart::l1_origin`]. Downstream
    /// consumers and the DA payload see the same origin the sealer set.
    pub l1_origin: u64,
}

File: cluster/sealer-service/core/src/main/java/io/kardamom/sealer/Boundary.java
package io.kardamom.sealer;

/**
 * A block boundary stamped at a tick. This is an immutable value class.
 *
 * <p>This class is a port of Rust's {@code BlockBoundaryStart}. The sealer's
 * boundary emitter produces it (see {@code crates/sealer/src/emitter.rs}).</p>
 * <ul>
 *   <li>{@code blockNumber} — the block that this boundary opens.</li>
 *   <li>{@code endTxIdx} — the cumulative count of canonical records published
 *       before this tick (mirrors {@code end_tx_idx = BPosition::from_index(canonical_count)}).</li>
 *   <li>{@code l2Timestamp} — the leader clock, floored to the 250 ms tick
 *       interval and forced strictly greater than the previous boundary's.</li>
 *   <li>{@code l1Origin} — the L1 block number for this block's epoch. The
 *       sealer never reads L1. It echoes the value from the last ordered
 *       origin-advancing record (see
 *       {@code docs/agents/l1-origin-deposit-derivation-spec.md}). The value
 *       is {@code 0} until the first such record is ordered.</li>
 * </ul>
 */
public final class Boundary {
    public final long blockNumber;
    public final long endTxIdx;
    public final long l2Timestamp;
    public final long l1Origin;

    public Boundary(long blockNumber, long endTxIdx, long l2Timestamp, long l1Origin) {
        this.blockNumber = blockNumber;
        this.endTxIdx = endTxIdx;
        this.l2Timestamp = l2Timestamp;
        this.l1Origin = l1Origin;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Boundary)) {
            return false;
        }
        Boundary other = (Boundary) o;
        return blockNumber == other.blockNumber
                && endTxIdx == other.endTxIdx
                && l2Timestamp == other.l2Timestamp
                && l1Origin == other.l1Origin;
    }

    @Override
    public int hashCode() {
        int result = Long.hashCode(blockNumber);
        result = 31 * result + Long.hashCode(endTxIdx);
        result = 31 * result + Long.hashCode(l2Timestamp);
        result = 31 * result + Long.hashCode(l1Origin);
        return result;
    }

    @Override
    public String toString() {
        return "Boundary{blockNumber=" + blockNumber
                + ", endTxIdx=" + endTxIdx
                + ", l2Timestamp=" + l2Timestamp
                + ", l1Origin=" + l1Origin + '}';
    }
}
